harness/determined/common/api/_session.py

```diff
@@ -100,13 +159,103 @@ def put(
     ) -> requests.Response:
         return self._do_request("PUT", path, params, json, data, headers, timeout, False)
 
+
+class UnauthSession(BaseSession):
+    """
+    UnauthSession is mostly only useful to log in or unathenticated endpoints like /info.
+    """
+
+    def __init__(
+        self,
+        master: Optional[str],
+        cert: Optional[certs.Cert],
+        max_retries: Optional[urllib3.util.retry.Retry] = None,
+    ) -> None:
+        self._master = master or util.get_default_master_address()
+        self._cert = cert
+        self._max_retries = max_retries
+
+    def _do_request(
+        self,
+        method: str,
+        path: str,
+        params: Optional[Dict[str, Any]],
+        json: Any,
+        data: Optional[str],
+        headers: Optional[Dict[str, Any]],
+        timeout: Optional[int],
+        stream: bool,
+    ) -> requests.Response:
+        return _do_request(
+            method=method,
+            host=self._master,
+            path=path,
+            params=params,
+            json=json,
+            data=data,
+            headers=headers,
+            utp=None,
+            cert=self._cert,
+            timeout=timeout,
+            stream=stream,
+            max_retries=self._max_retries,
+        )
+
+
+class Session(BaseSession):
+    """
+    Session authenticates every request it makes.
+
+    By far, most BaseSessions in the codebase will be this Session subclass.
+    """
+
+    def __init__(
+        self,
+        master: Optional[str],
+        utp: authentication.UsernameTokenPair,
+        cert: Optional[certs.Cert],
+        max_retries: Optional[urllib3.util.retry.Retry] = None,
+    ) -> None:
+        self._master = master or util.get_default_master_address()
+        self._utp = utp
```

Comment:
I know the code's not set, but note here in case you meant to make this public.

Reply:
Or even better, give Session accessors to `username` and `token` directly.

Reply:
There are so few places where somebody should have access to username or token, I'd prefer leaving this as-is.

Reply:
I'd guess, there's like 2-3 thousand instances of Session in our system, and probably fewer than 5 that need the username.  You're right that it's a leaking abstraction... but at that level, I don't think I care a lot.

If you really feel strongly, we can expose the username maybe?  But I don't see any reason why somebody should need the token.

Another aspect here is that a leaky abstraction in an internal-facing object is not that big of a deal to me, at least not on this magnitude of leakage.

(technically the symbol is exposed in experimental.client but I might actually remove that when we land this change, since for other reasons this change will be a breaking change)

Reply:
The more I look at this, the more I agree that the Session should expose username and token.

Especially if it is explicitly removed from the `client` namespace, so that it is explicitly an internal struct.  I think some of the e2e tests could also be simplified a bit in that case.

Side note, there's fewer than 2-3 thousand instances of Session, prolly closer to 700-800... I git-grep'd.  And there's 15 instances of somebody accessing `._utp` on a session in the codebase.

Reply:
The more I thought about it, the more I thought that `.master` and `.cert` should also be accessible.  If the Session is supposed to be a reusable collection of master url, cert, username, and token... then we might as well make it easy to reuse those elements.